src/stm32_generic/i2c.cpp

```diff
@@ -0,0 +1,548 @@
+  // register. bit_modify(i2c_reg->cr1).set(i2c_cr1::peripheral_enable);
+
+  // Set the SDA line to high, so that after the slave releases
+  // the SDA and listen for ACK/NACK, it will get a NACK.
+  sda->level(true);
+
+  for (int i = 0; i < 10; ++i) {
+    // Set SCL high.
+    scl->level(true);
+    continue;
+    // Set SCL low.
+    scl->level(false);
+    continue;
+  }
+  // Set SCL high. When I2C is idle the line should be pulled high.
+  scl->level(true);
+}
+
+void i2c::configure(hal::i2c::settings const& p_settings, hertz p_frequency)
+{
+  constexpr auto slow_mode_max_speed = 100_Hz;
+  if (static_cast<u32>(p_frequency) < 2_MHz ||
+      static_cast<u32>(p_frequency) > 50_MHz) {
+    safe_throw(hal::argument_out_of_domain(this));
+  }
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  bit_modify(i2c_reg->cr1).set(i2c_cr1::software_reset);
+  bit_modify(i2c_reg->cr1).clear(i2c_cr1::software_reset);
+
+  bit_modify(i2c_reg->filter)
+    .set(i2c_filter::digital_filter)
+    .clear(i2c_filter::analog);
+
+  /// I2C communication speed, fSCL ~ 1/(thigh + tlow). The real frequency may
+  /// differ due to the analog noise filter input delay.
+  /// CH 18.6.8 in RM0383 (stm32f411 user manual)
+
+  u16 const ccr_value = static_cast<hal::u32>(p_frequency) /
+                        (2 * static_cast<hal::u32>(p_settings.clock_rate));
+  if (p_settings.clock_rate > slow_mode_max_speed) {
+    if (ccr_value > 4) {
+      bit_modify(i2c_reg->ccr)
+        .set(i2c_ccr::i2c_mode_sel)
+        .insert(i2c_ccr::ccr, ccr_value);
+    } else {
+      safe_throw(hal::argument_out_of_domain(this));
+    }
+  } else {
+    bit_modify(i2c_reg->ccr)
+      .clear(i2c_ccr::i2c_mode_sel)
+      .insert(i2c_ccr::ccr, ccr_value);
+  }
+
+  auto const perf_freq = static_cast<hal::u8>(p_frequency / 1_MHz);
+  bit_modify(i2c_reg->cr1)
+    .set(i2c_cr1::peripheral_enable)
+    .set(i2c_cr1::ack_enable)
+    .clear(i2c_cr1::smbus_mode);
+  bit_modify(i2c_reg->cr2)
+    .set(i2c_cr2::buffer_intr_en)
+    .set(i2c_cr2::event_intr_en)
+    .set(i2c_cr2::error_intr_en)
+    .insert(i2c_cr2::apb_frequency, perf_freq);
+}
+
+void i2c::handle_i2c_event() noexcept
+{
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  auto& status = i2c_reg->sr1;
+  auto& data = i2c_reg->data_reg;
+  if (bit_extract<i2c_sr1::start>(status)) {
+    if (!m_data_out.empty()) {
+      data = to_8_bit_address(m_address, i2c_operation::write);
+      m_transmitter = true;
+      m_reciever = false;
+    } else {
+      data = to_8_bit_address(m_address, i2c_operation::read);
+      m_transmitter = false;
+      m_reciever = true;
+    }
+    return;
+  }
+
+  if (bit_extract<i2c_sr1::addr>(status)) {
+    if (m_data_in.size() == 1 && m_reciever) {
+      bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable);
+    }
+    if (m_data_in.size() == 2 && m_reciever) {
+      bit_modify(i2c_reg->cr1).clear(i2c_cr1::ack_enable).set(i2c_cr1::pos);
+    }
+    [[maybe_unused]] auto addr = i2c_reg->sr2;
```

Comment:
```suggestion
    [[maybe_unused]] volatile auto addr = i2c_reg->sr2;
```

Add `volatile` to make absolutely sure that the object gets loaded with a value, otherwise, if this gets optimized away for some reason, then this code may get skipped as it doesn't cause a side effect, and then your code does not work.

Reply:
Technically the sr2 being volatile should also do this, but I think its better to have them volatile on both sides.

Reply:
when you say "sr2 being volatile" is this something coded into the hardware and will always act volatile?

Reply:
In the struct, sr2 should be marked as volatile. If it isn't then it should be. But basically, we want to ensure that the read happens because if the compiler decides that the operation does nothing, it may still optimize it away. I think that is something important to look into as a style guide requirement. Can you investigate if a statement like `i2c_reg->sr2` by itself will always perform a read operation on that register? Do you always need to have a variable it gets stored into with `[[maybe_used]]`?